batched/dense/impl/KokkosBatched_Pbtrf_Serial_Internal.hpp

```diff
@@ -0,0 +1,268 @@
+#define KOKKOSBATCHED_PBTRF_SERIAL_INTERNAL_HPP_
+
+#include "KokkosBatched_Util.hpp"
+#include "KokkosBlas1_serial_scal_impl.hpp"
+
+namespace KokkosBatched {
+
+///
+/// Serial Internal Impl
+/// ====================
+
+///
+/// Lower
+///
+
+template <typename AlgoType>
+struct SerialPbtrfInternalLower {
+  template <typename ValueType>
+  KOKKOS_INLINE_FUNCTION static int invoke(const int an,
+                                           /**/ ValueType *KOKKOS_RESTRICT AB, const int as0, const int as1,
+                                           const int kd);
+
+  template <typename ValueType>
+  KOKKOS_INLINE_FUNCTION static int invoke(const int an,
+                                           /**/ Kokkos::complex<ValueType> *KOKKOS_RESTRICT AB, const int as0,
+                                           const int as1, const int kd);
+};
+
+///
+/// Real matrix
+///
+
+template <>
+template <typename ValueType>
+KOKKOS_INLINE_FUNCTION int SerialPbtrfInternalLower<Algo::Pbtrf::Unblocked>::invoke(const int an,
+                                                                                    /**/ ValueType *KOKKOS_RESTRICT AB,
+                                                                                    const int as0, const int as1,
+                                                                                    const int kd) {
+  // Compute the Cholesky factorization A = L*L'.
+  for (int j = 0; j < an; ++j) {
+    auto a_jj = AB[0 * as0 + j * as1];
+
+    // Check if L (j, j) is positive definite
+#if (KOKKOSKERNELS_DEBUG_LEVEL > 0)
+    if (a_jj <= 0) {
+      return j + 1;
+    }
+#endif
+
+    a_jj                  = Kokkos::sqrt(a_jj);
+    AB[0 * as0 + j * as1] = a_jj;
+
+    // Compute elements J+1:J+KN of column J and update the
+    // trailing submatrix within the band.
+    int kn = Kokkos::min(an - j - 1, kd);
+    if (kn > 0) {
+      // scale to diagonal elements
+      const ValueType alpha = 1.0 / a_jj;
+      KokkosBlas::Impl::SerialScaleInternal::invoke(kn, alpha, &(AB[1 * as0 + j * as1]), 1);
+
+      // syr (lower) with alpha = -1.0 to diagonal elements
+      for (int k = 0; k < kn; ++k) {
+        auto x_k = AB[(k + 1) * as0 + j * as1];
+        if (x_k != 0) {
+          auto temp = -1.0 * x_k;
+          for (int i = k; i < kn; ++i) {
+            auto x_i = AB[(i + 1) * as0 + j * as1];
+            AB[i * as0 + (j + 1 + k - i) * as1] += x_i * temp;
+          }
+        }
+      }
+    }
+  }
+
+  return 0;
+}
+
+///
+/// Complex matrix
+///
+template <>
+template <typename ValueType>
+KOKKOS_INLINE_FUNCTION int SerialPbtrfInternalLower<Algo::Pbtrf::Unblocked>::invoke(
+    const int an,
+    /**/ Kokkos::complex<ValueType> *KOKKOS_RESTRICT AB, const int as0, const int as1, const int kd) {
+  // Compute the Cholesky factorization A = L*L**H
+  for (int j = 0; j < an; ++j) {
+    auto a_jj = AB[0 * as0 + j * as1].real();
+
+    // Check if L (j, j) is positive definite
+    if (a_jj <= 0) {
```

Comment:
This is actually on purpose, based on the [reference implementation](https://www.netlib.org/lapack/explore-html/d4/d16/group__pbtf2_ga740a2a60e8a6cf6f397e6d4e2070df54.html#ga740a2a60e8a6cf6f397e6d4e2070df54). After some consideration, I added a debug level guard. In case `L (j, j)` is not positive definite, subsequent operations are invalid anyway. We do not care whether the complex part of `L (j, j)` is suppressed or not for the invalid case.